### ISSUE TITLE
Connect Apache Beam/Spark to TensorFlow (MonitoredTrainingSession) in a streaming manner?

### ISSUE THREAD

**Issue by MtDersvan** [tensorflow/tensorflow#16082]
### Describe the problem
I have a lengthy question on [SO](https://stackoverflow.com/questions/47986410/optimal-data-streaming-and-processing-solution-for-enormous-datasets-into-tf-dat) about this. But in short, is there a way (or a best practice) to pipe big training datasets directly into a distributed setting (e.g. GKE),  especially if they are subjected to a heavy preprocessing? 
I'm basically reaching the limit of what can be sanely stored in TFRecords (they are verbose and heavy).
The closest issue was this one (https://github.com/tensorflow/tensorflow/issues/12903) and this guide (https://github.com/GoogleCloudPlatform/dataflow-prediction-example) but I do not see a healthy way to implement it (last one with a `@singleton` looks like a hack and not usable with the `tf.Dataset` or `MonitoredTrainingSession`).

I believe this is a useful issue/feature request for a decent amount of Tensorflow users.

**Comment by yongtang** [comment]
There is a pending PR #14098 that adds a KafkaReader that might be related.

**Comment by MtDersvan** [comment]
@yongtang Yes! I'm subscribed to that PR, and already looked at the kernel (and it looks nice, well done!). Kafka would be a nice option to have.
The infrastructure, though, is scattered, and not well documented. For example, Is there a way to stream data directly from Apache Beam/Spark (or even Kafka) to a `tf.data.Dataset_from_generator()` without dumping data into `tfrecords` files? 
Maybe just a simple streaming serialization support a la Tensorflow Serving's `predict_pb2.PredictRequest()`  +  `CopyFrom(make_tensor_proto(SerializeToString())` can do the trick, as it works really well? (it can still be a `tf.Example`, `tf.SequenceExample` proto but in a more convenient online way)

**Comment by poxvoculi** [comment]
Adding @mrry for thoughts on best practices in this case.

**Comment by mrry** [comment]
If you're using PySpark and wrapping your training step in an `RDD.mapPartitions()` transformation, you should be able to use `tf.data.Dataset.from_generator()` to convert the partition iterator into a generator of NumPy arrays. I dare say there might be a more efficient way to connect a Spark RDD to a `tf.data.Dataset` (e.g. with a custom reader dataset as @yongtang suggested), and we'd welcome contributions of that kind.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?
